cli/mender-artifact/artifacts.go

```diff
@@ -199,231 +169,184 @@ func repack(comp artifact.Compressor, artifactName string, from io.Reader, to io
-		ar.RegisterHandler(rootfs)
+	err = aReader.ReadArtifactHeaders()
+	if err != nil {
+		return nil, err
 	}
 
-	r, err := read(ar, verify, storeScripts)
+	fDir := filepath.Join(tmpdir, "files")
+	err = os.Mkdir(fDir, 0755)
 	if err != nil {
 		return nil, err
 	}
 
-	info := r.GetInfo()
-
-	// now once arifact is read we need to
-	var h *handlers.Rootfs
-	switch info.Version {
-	case 1:
-		return nil, errors.New("Mender-Artifact version 1 no longer supported")
-	case 2:
-		h = handlers.NewRootfsV2(data)
-	case 3:
-		h = handlers.NewRootfsV3(data)
-	default:
-		return nil, errors.Errorf("unsupported artifact version: %d", info.Version)
+	updateStore := &writeUpdateStorer{
+		dir: fDir,
 	}
-
-	upd := &awriter.Updates{
-		Updates: []handlers.Composer{h},
+	inst := aReader.GetHandlers()
+	if len(inst) == 1 {
+		inst[0].SetUpdateStorerProducer(updateStore)
+	} else if len(inst) > 1 {
+		return nil, errors.New("More than one payload not supported")
 	}
-	scr, err := scripts([]string{sDir})
-	if err != nil {
+
+	if err := aReader.ReadArtifactData(); err != nil {
 		return nil, err
 	}
 
-	aWriter := awriter.NewWriter(to, comp)
-	if key != nil {
-		aWriter = awriter.NewWriterSigned(to, comp, artifact.NewSigner(key))
-	}
+	ua.files = updateStore.names
 
-	name := ar.GetArtifactName()
-	provides := ar.GetArtifactProvides()
-	if newName != "" {
-		name = newName
-		if provides != nil {
-			provides.ArtifactName = newName
-		}
-	}
+	if len(inst) > 0 &&
+		inst[0].GetUpdateType() == "rootfs-image" &&
+		len(ua.files) != 1 {
 
-	typeInfoV3 := artifact.TypeInfoV3{
-		Type: "rootfs-image",
-		// Keeping these empty for now. We will likely introduce these
-		// later, when we add support for augmented artifacts.
-		// ArtifactDepends:  &artifact.TypeInfoDepends{"rootfs_image_checksum": c.String("depends-rootfs-image-checksum")},
-		// ArtifactProvides: &artifact.TypeInfoProvides{"rootfs_image_checksum": c.String("provides-rootfs-image-checksum")},
-		ArtifactDepends:  &artifact.TypeInfoDepends{},
-		ArtifactProvides: &artifact.TypeInfoProvides{},
+		return nil, errors.New("rootfs-image artifacts with more than one file not supported")
 	}
 
-	err = aWriter.WriteArtifact(
-		&awriter.WriteArtifactArgs{
-			Format:     info.Format,
-			Version:    info.Version,
-			Devices:    ar.GetCompatibleDevices(),
-			Name:       name,
-			Updates:    upd,
-			Scripts:    scr,
-			Provides:   provides,
-			Depends:    ar.GetArtifactDepends(),
-			TypeInfoV3: &typeInfoV3,
-		})
-	return ar, err
+	ua.writeArgs, err = reconstructArtifactWriteData(ua)
+	return ua, err
 }
 
-func repackArtifact(comp artifact.Compressor, artifact, rootfs, newName string) error {
-	art, err := os.Open(artifact)
-	if err != nil {
-		return err
-	}
-	defer art.Close()
+func reconstructPayloadWriteData(info *artifact.Info, inst map[int]handlers.Installer) (upd *awriter.Updates,
+	typeInfoV3 *artifact.TypeInfoV3,
+	augTypeInfoV3 *artifact.TypeInfoV3,
+	metaData map[string]interface{},
```

Comment:
Suggestion: Wow, lots of return values. Wrap these in a struct ?  Basically this is half an artifact worth of information right? I don't see why this cannot return the `writeArtifactArgs` struct ?

Reply:
I like to follow the "principle of least information", which is why I don't want it to return the whole struct. It's lacking many fields from that struct, so it is incomplete.

I could turn it into its own struct, but this is only used in one single place, so I didn't really see much point. You'd just have all the same definitions right above the function instead of in the function. I think my vote would be for keeping it as is. Do you think it looks too messy?

Reply:
Yeh, I figured that was your rational. It is fine with me, it's not too messy. I would have been more stanch if this was parameters accepted and not returned I think ;)